### ISSUE TITLE
[NCCL-Hang] ALLREDUCE timeout at SeqNum=3 after clip_grad_norm — optimizer.step never reached

### ISSUE THREAD

**Issue by dylanyunlon** [dylanyunlon/Neuron_SP#197]
## Symptom
NCCL watchdog timeout (30min) on ALLREDUCE with NumelIn=1 after `clip_grad_norm` completes.

## Timeline from logs
```
04:04:46 rank=0: EXITED clip_grad_norm gnorm=3861.279541
... 30 minutes of silence ...
04:34:46 WorkNCCL(SeqNum=3, OpType=ALLREDUCE, NumelIn=1) TIMEOUT
```

## Root cause hypothesis
`finalize_model_grads` has `skip=True` but `force_ar=True` — this means rank 0 issues an allreduce but other ranks may not (conditional collective). SeqNum=3 means only 3 NCCL ops were issued before hang, suggesting the mismatch happens very early in the gradient finalization path.

## Key code path
```
desloc_engine.py → finalize_model_grads(Kx=False, skip=True, force_ar=True)
→ some allreduce on a scalar (NumelIn=1, likely grad_norm or nan_flag)
→ rank 0 participates, rank 1/2 do not → HANG
```

## Fix approach
1. Audit every `dist.all_reduce` in the path between `clip_grad_norm` exit and `optimizer.step` entry
2. Ensure all conditional collectives use the SAME condition on ALL ranks
3. The `NumelIn=1` allreduce is likely a nan-check or grad-norm broadcast — make it unconditional

## Related commits (all attempted fixes that did not resolve)
- 8de1af22 fix(desloc_engine): fix post-clip hang
- ff53fc7f fix(grad_norm): fix remaining CPU tensor collectives in clip path
- 3f939bba fix(grad_norm): ensure all dist.all_reduce tensors are on CUDA
- e07c5a7f fix(hetero_grad_norm_skip): move partial_tensor to CUDA before allreduce
- f02096e2 fix(desloc_engine): make NaN check local-only
- 0dfd3628 fix(desloc_engine): remove runtime num_microbatches allreduce

## Environment
- 3 GPU: A6000 (rank0) + A6000 (rank1) + H100 NVL (rank2)
- PCIe only, no NVLink, cross-NUMA
- PyTorch 2.7+ with CUDA 13.0
- NCCL timeout: 1800s
